src/P2P/Node.hs

```diff
@@ -298,39 +296,87 @@ setInactive :: P2pNode -> STM ()
 setInactive node = writeTVar (_p2pNodeActive node) False
 
 -- -------------------------------------------------------------------------- --
--- Sync Peers
+-- New Peer Validation Manager
+
+-- | Global Manager for checking reachability of new Peers
+--
+newPeerManager :: IORef HTTP.Manager
+newPeerManager = unsafePerformIO
+    $ unsafeManager 2000000 {- 1 second -} >>= newIORef
+{-# NOINLINE newPeerManager #-}
+
+getNewPeerManager :: IO HTTP.Manager
+getNewPeerManager = readIORef newPeerManager
+{-# INLINE getNewPeerManager #-}
+
+-- -------------------------------------------------------------------------- --
+-- Guard PeerDB
+
+data PeerValidationFailure
+    = IsReservedHostAddress !PeerInfo
+    | IsNotReachable !PeerInfo !T.Text
+    deriving (Show, Eq, Ord, Generic, NFData, ToJSON)
+
+instance Exception PeerValidationFailure where
+    displayException (IsReservedHostAddress p)
+        = "The peer info " <> T.unpack (showInfo p) <> " is form a reserved IP address range"
+    displayException (IsNotReachable p t)
+        = "The peer info " <> T.unpack (showInfo p) <> " can't be reached: " <> T.unpack t
 
 -- | Removes candidate `PeerInfo` that are:
 --
 --  * already known to us and considered good
 --  * are trivially bad (localhost, our own current IP, etc.)
 --
-removeTrivialPeers :: P2pNode -> PeerInfo -> IO (Maybe PeerInfo)
-removeTrivialPeers node pinf = do
-    peers <- peerDbSnapshot $ _p2pNodePeerDb node
-    if | me == _peerId pinf || isTrivial || isKnown peers -> pure Nothing
-       | otherwise -> bool Nothing (Just pinf) <$> canConnect
+guardPeerDb
+    :: ChainwebVersion
+    -> NetworkId
+    -> PeerDb
+    -> PeerInfo
+    -> IO (Either PeerValidationFailure PeerInfo)
+guardPeerDb v nid peerDb pinf = do
+    peers <- peerDbSnapshot peerDb
+    if
+        | isKnown peers -> return $ Right pinf
+        | isReserved -> return $ Left $ IsReservedHostAddress pinf
+        | otherwise -> canConnect >>= \case
+            Left e -> return $ Left $ IsNotReachable pinf (sshow e)
+            Right _ -> return $ Right pinf
   where
-    v = _chainwebVersion node
-    env = peerClientEnv node pinf
-    nid = _p2pNodeNetworkId node
-
-    me :: Maybe PeerId
-    me = _peerId $ _p2pNodePeerInfo node
-
     isKnown :: PeerSet -> Bool
     isKnown peers = not . IXS.null $ IXS.getEQ (_peerAddr pinf) peers
 
     -- TODO Check more IP ranges, possibly in HostAddress module.
-    isTrivial :: Bool
-    isTrivial = case v of
+    isReserved :: Bool
+    isReserved = case v of
         Mainnet01 -> isReservedHostAddress (_peerAddr pinf)
+        Testnet02 -> isReservedHostAddress (_peerAddr pinf)
+        Development -> isReservedHostAddress (_peerAddr pinf)
         _ -> False
 
-    canConnect :: IO Bool
-    canConnect = runClientM (peerGetClient v nid (Just 0) Nothing) env >>= \case
-        Left _ -> pure False
-        Right _ -> pure True
+    canConnect = do
+        mgr <- getNewPeerManager
```

Comment:
This is just calling the `Manager` that was already created once, not generating a new `Manager` each time? The latter would be quite expensive.